src/nvim/os/env.c

```diff
@@ -169,6 +178,64 @@ void os_get_hostname(char *hostname, size_t size)
 #endif
 }
 
+/// @return -1 if error encountered, 1 if capslock active otherwise 0 if not.
+int capslock_status(void ) {
+
+#if defined(__linux__)
+
+  FILE *fp = NULL;
+  char res[1024];
+  fp = popen("xset -q | grep Caps", "r");
+  if (fp == NULL) {
+      return -1;
+  }
+  fgets(res, sizeof(res), fp);
+  if (STRNCMP(res + 21, "on", 2) == 0) {
+    return 1;
+  } else {
+    return 0;
+  }
+
+#elif defined(WIN32)
+  if ((GetKeyState(VK_CAPITAL) & 0x0001)!=0) {
+    return 1
+  } else {
+    return 0;
+  }
+#elif defined(__APPLE__)
+
+  kern_return_t kr;
+  io_service_t ios;
+  io_connect_t ioc;
+  CFMutableDictionaryRef mdict;
+  bool state;
+  mdict = IOServiceMatching(kIOHIDSystemClass);
+  ios = IOServiceGetMatchingService(kIOMasterPortDefault, (CFDictionaryRef) mdict);
+  if (!ios) {
+    if (mdict) {
+      CFRelease(mdict);
```

Comment:
Does `mdict` also need to be released when `ios` is valid?

Reply:
Yes, I saw the documentation of `IOServiceGetMatchingService`, they say, if there is an error encountered then it will not release the dictionary, otherwise it will release the space